File: springapp/src/main/java/com/examly/springapp/service/VehicleDetailsService.java
package com.examly.springapp.service;

import java.util.List;

import com.examly.springapp.dto.VehicleDto;

public interface VehicleDetailsService {

    VehicleDto createVehicle(VehicleDto vehicleDto);
    VehicleDto updateVehicle(Integer id, VehicleDto vehicleDto);
    VehicleDto getVehicleById(Integer id);
    List<VehicleDto> getAllVehicles();
    void deleteVehicle(Integer id);
    List<VehicleDto> searchByBrandName(String brandName);
    List<VehicleDto> filterByVehicleType(String vehicleType);
    List<VehicleDto> filterByFuelType(String fuelType);

}

File: angularapp/src/app/services/vehicle-details.service.ts
import { Injectable } from '@angular/core';
import { Vehicle } from '../models/vehicle.model';
import { API_URL } from '../constants/app.constant';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';

@Injectable({
  providedIn: 'root'
})
export class VehicleDetailsService {

  private selectedVehicleId: number | null = null;
  
  constructor(private http: HttpClient) {}
  
  setSelectedVehicleId(id: number): void {
    this.selectedVehicleId = id;
  }

  getSelectedVehicleId(): number | null {
    return this.selectedVehicleId;
  }


  public getAllVehicles(): Observable<Vehicle[]> {
    return this.http.get<Vehicle[]>(API_URL + '/vehicles');
  }

  public createVehicle(vehicle: Vehicle): Observable<Vehicle> {
    return this.http.post<Vehicle>(API_URL + '/vehicles', vehicle);
  }

  public updateVehicle(vehicleId: number, updatedVehicle: Vehicle): Observable<Vehicle> {
    return this.http.put<Vehicle>(API_URL + '/vehicles/' + vehicleId, updatedVehicle);
  }

  public deleteVehicle(vehicleId: number): Observable<any> {
    return this.http.delete<any>(API_URL + '/vehicles/' + vehicleId);
  }

  public getVehicleById(vehicleId: number): Observable<Vehicle> {
    return this.http.get<Vehicle>(API_URL + '/vehicles/' + vehicleId);
  }

  public searchByBrand(brandName: string): Observable<Vehicle[]> {
    return this.http.get<Vehicle[]>(API_URL + '/vehicles/brand/' + brandName);
  }

  public filterByType(vehicleType: string): Observable<Vehicle[]> {
    return this.http.get<Vehicle[]>(API_URL + '/vehicles/type/' + vehicleType);
  }

  public filterByFuel(fuelType: string): Observable<Vehicle[]> {
    return this.http.get<Vehicle[]>(API_URL + '/vehicles/fuel/' + fuelType);
  }

}
